### PR TITLE
Experimentation with MLM pretraining

### PR DESCRIPTION
**Pull request by LouisCastricato**
We're conducting experiments in using an MLM objective to improve data efficiency.

This PR will include:

A new data pipeline API for collating and processing input data.
A new learning rate scheduler API for custom LR schedulers that are model state aware.
CARP MLM, which optimizes MLM for a fixed number of epochs before switching to a different CARP objective.
MixedMLMEncoder, which will start as an MLM encoder and switch to some other encoder after a fixed number of epochs.

### CODE REVIEW

**Comment by LouisCastricato** [comment]
Data pipeline ready 2 roll

**Comment by LouisCastricato** [comment]
While this PR was primarily focused on adding MLM pretraining to magiCARP, in the process a large rearchitecture was needed.


Tokenizer factories are now abstracted behind data pipeline classes. The implementation is sexy as fuck, but the variable names are a little bit confusing (On part of the API that would not be used by developers however). 

See here: 
https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/data/__init__.py#L73


Instead of tensors being passed through the data pipeline and model, communication between the data pipeline and model is done with an instance of "BatchElement"

https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/data/utils/data_util.py#L85

And communication from the encoders to model is done using an instance of "BaseEncoderOutput." Both still allow for static typing.

https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/model/encoders/__init__.py#L51

The model class now handles gradient accumulation. 

https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/model/architectures/__init__.py#L153

Autocast will also be handled here in the future to allow for better interopability between AMP and DeepSpeed

CARP Momentum needs to be retested and potentially abstracted a bit for alternatively momentum methods.

https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/model/architectures/carp_momentum.py

MLM encoder has a really odd memory leak. For some reason increasing batch size but leaving microbatch size constant leads to OOM- for instance training MLM on a batch size of 1024 but microbatch of 1 leads to OOM on a 3090. This might be because it is storing too much gradient information (?) It wants to go through all the microbatches before it begins calling loss.backward() (Or AMP's equivalent) and that might be leading to issues. As of right now we are stuck with a batch size of 64. @shahbuland I need you to fix this. 

https://github.com/EleutherAI/magiCARP/blob/485a1f6af09a2ea8194b99bfd84168ba6c5066bb/carp/pytorch/model/architectures/carp_mlm.py#L119

Finally I added the orchestrator which allows for calls to be added to the training/validation/testing loop at various points. It seems interesting for odd scheduling functions, and it was required for CARP MLM- which runs its first epoch as an MLM model and future epochs as a contrastive learning model.

**Comment by LouisCastricato** [comment]
Memory leak is mostly fixed, the validation loss step on contrastive learning for some reason crashes lol- do not understand why.

File: carp/pytorch/model/architectures/carp_mlm.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+from carp.configs import CARPConfig, ModelConfig, TrainConfig
+from carp.pytorch.data.mlm_pipeline import MLMBatchElement
+from carp.pytorch.model.architectures import * 
+from carp.pytorch.model.encoders import get_encoder
+from carp.util import mbTokens, generate_indices
+from typing import List
+
+
+
+# CARP MLM differs from normal CARP since the first epoch will solely use an MLM objective to improve data efficiency.
```
Comment (dashstander) [inline carp/pytorch/model/architectures/carp_mlm.py:15]: Only the first epoch?

Comment (dashstander) [reply carp/pytorch/model/architectures/carp_mlm.py:15]: Can it not just be an auxiliary loss the entire time?

File: carp/pytorch/tests/test_dataset.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import pytest
```
Comment (dashstander) [inline carp/pytorch/tests/test_dataset.py:1]: Do we need both `carp/pytorch/tests/test_dataset.py` and `carp/pytorch/test_dataset.py`?

File: carp/pytorch/model/architectures/carp_mlm.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+
+        self.mlm_mode = True
+
+    def _embed_data(
+        self,
+        x: MLMBatchElement,
+        encoder,
+        projector,
+    ):
+        if self.mlm_mode:
```
Comment (dashstander) [inline carp/pytorch/model/architectures/carp_mlm.py:50]: Could we instead just pass in `mlm_mode` as an argument to the function instead of maintaining the state in the orchestrator? Factoring out the orchestrator is clean, but it doesn't do much and it makes the logic a little convoluted--I had to go back and forth and check a few different files to understand where this was getting controlled. Just having a function in the training script would be less elegant, but it'd be easier to read/understand. As things get added we'll, in all likelihood, want something along the lines of an "orchestrator", but I don't think things are quite that complicated yet.

**Review by dashstander — COMMENTED** [review]
Things look pretty good. I only have the one small concern about the orchestrator class (that I explained in a comment), but I haven't thought through exactly what an alternative would look like, so if you think that it's the best option go for it.

**Comment by shahbuland** [comment]
Found the cause of OOM issue on eval_step. 
https://github.com/EleutherAI/magiCARP/blob/b28f8d7782146fc8553f2c2adbe90c0c28ee2fb3/carp/pytorch/model/architectures/__init__.py#L175  
This calls here
https://github.com/EleutherAI/magiCARP/blob/b28f8d7782146fc8553f2c2adbe90c0c28ee2fb3/carp/pytorch/model/architectures/carp_mlm.py#L64-L88   
Checking passages before line 79 shows it is a list of length one. The first element is of this type
https://github.com/EleutherAI/magiCARP/blob/b28f8d7782146fc8553f2c2adbe90c0c28ee2fb3/carp/pytorch/data/mlm_pipeline.py#L11-L13
and the input_ids field has shape (1000, 166). This means that  
https://github.com/EleutherAI/magiCARP/blob/b28f8d7782146fc8553f2c2adbe90c0c28ee2fb3/carp/pytorch/model/architectures/carp_mlm.py#L80
Is trying to do a forward pass through model with a batch of size 1000. This is why it keeps OOMing. Additionally MLMBatchElement doesn't seem to be iterable so `for p in passages[0]` would not work either. I have two potential solutions and want to know which you prefer.  
1. Add \_\_getitem\_\_ to BatchElement class to make it iterable.
2. Make a utility function that splits a BatchElement class into multiple smaller BatchElement objects such that each has at most some size tensors in it.
